designs/extensible-cli-and-scaffolding-plugins-phase-1-5.md

```diff
@@ -0,0 +1,286 @@
+# Extensible CLI and Scaffolding Plugins - Phase 1.5
+
+Continuation of [Extensible CLI and Scaffolding Plugins](./extensible-cli-and-scaffolding-plugins-phase-1.md).
+
+<a id="goal" />
+
+## Goal
+
+The goal of this phase is to achieve one of the goals proposed for Phase 2: chaining plugins.
+Phase 2 includes several other challenging goals, but being able to chain plugins will be beneficial
+for third-party developers that are using kubebuilder as a library.
+
+## Table of contents
+- [Goal](#goal)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+
+<a id="motivation" />
+
+## Motivation
+
+There are several cases of plugins that want to maintain most of the go plugin functionality and add
+certain features on top of it, both inside and outside kubebuilder repository:
+- [Addon pattern](../plugins/addon)
+- [Operator SDK](https://github.com/operator-framework/operator-sdk/tree/master/internal/plugins/golang)
+
+This behavior fits perfectly under Phase 1.5, where plugins could be chained. However, as this feature is
+not available, the adopted temporal solution is to wrap the base go plugin and perform additional actions
+after its `Run` method has been executed. This solution faces several issues:
+
+- Wrapper plugins are unable to access the data of the wrapped plugins, as they weren't designed for this
+  purpose, and therefore, most of its internal data is non-exported. An example of this inaccessible data
+  would be the `Resource` objects created inside the `create api` and `create webhook` commands.
+- Wrapper plugins are dependent on their wrapped plugins, and therefore can't be used for other plugins.
+- Under the hood, subcommands implement a second hidden interface: `RunOptions`, which further accentuates
+  these issues.
+
+Plugin chaining solves the aforemention problems but the current plugin API, and more specifically the
+`Subcommand` interface, does not support plugin chaining.
+
+- The `RunOptions` interface implemented under the hood is not part of the plugin API, and therefore
+  the cli is not able to run post-scaffold logic (implemented in `RunOptions.PostScaffold` method) after
+  all the plugins have scaffolded their part.
+- `Resource`-related commands can't bind flags like `--group`, `--version` or `--kind` in each plugin,
+  it must be created outside the plugins and then injected into them similar to the approach followed
+  currently for `Config` objects.
+
+<a id="proposal" />
```

Comment:
FYI, you'll get autolinking from the heading titles, so you shouldn't need to add these in manually unless you don't want to have to worry about rewording the titles.

Reply:
As that is not markdown itself but GitHub markdown, I wasn't sure if it added them or not.